### ISSUE TITLE
[Bug] No FM sound in Shrines of Enigma [sf#82]

### ISSUE THREAD

**Issue by openMSX-import** [openMSX/openMSX#383]
**Reported by manuelbi on 2003-02-18 21:41 UTC**
The FM sound in &amp;quot;The Shrines of Enigma &amp;quot; that worked
just fine before, doesn't work anymore... There is no
FM sound at all anymore, only PSG. Still worked with
Gianna Sisters, e.g., though.

**Comment by openMSX-import** [comment]
**Commented by manuelbi on 2003-02-19 21:54 UTC**
Logged In: YES 
user_id=78178

First trying CALL MUSIC in basic and then booting doesn't
work: the game hangs then.

I tried with an adapted Panasonic FS-A1WSX config, and now I
do get FM sound.

However: Albert Beevendorp checked the code on the diskimage
and he says that the FM PAC is initialized in Shrines of
Enigma! 
Maybe something is wrong with the FMPAC on/off switching? Or
the initialization is not correct in the game?

**Comment by openMSX-import** [comment]
**Commented by manuelbi on 2003-02-20 21:54 UTC**
Logged In: YES 
user_id=78178

I asked someone to test it on a real MSX2. The same happens!
No FM sound. 
The game doesn't initialize the FM PAC at all. When booting
with DOS, typing CALL MUSIC, CALL SYSTEM and then from drive
B: booting the game with 'portal', the game *does* have FM
Sound on real MSX and exactly the same happens in openMSX. 

Also, Maarten wrote:

I added prints when the FMPAC memory range was read or
written. Only the BIOS 
accesses it: 0x4000-0x4009 for cartridge info and 0x4010 for
a write test to 
check whether this is RAM. The game never accesses the FMPAC
memory.

I investigated the disk image as well. Indeed the search and
initialisation 
code is there, but it is never called. I put a &amp;quot;DI ; HALT&amp;quot;
in the search code 
and the game worked just fine. It seems they hardcoded the
result as 
&amp;quot;MSX-MUSIC found&amp;quot; and didn't bother to run the check
anymore, forgetting that 
the initialisation side effect is necessary for Panasonic
FMPACs.

The variable at offset 0x4C09 in the disk image contains the
sound chip found. 
It is initially 1, which means MSX-MUSIC. The first thing
the search routine 
does is set this variable to 0. If you set it to 0 using a
hex editor, then 
an internal MSX-MUSIC does not give any sound either. If a
search routine is 
used, I would expect this variable to be 0 initially, but
they made it 1 
instead, probably because they knew they weren't calling the
search routine.

Maybe it was a hack to enable PSG sounds during the intro.
If there is no 
sound chip found, the PSG sounds are disabled as well. By
skipping the search 
and hardcoding the chip to MSX-MUSIC, the PSG sounds are
always enabled, even 
if you don't have an MSX-MUSIC. There are writes to the
MSX-MUSIC I/O ports 
as well, but that doesn't hurt if you don't have one.

All in all: this is no bug! openMSX did just fine!
